src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -151,41 +143,151 @@ private Task ConsumeChunkBytesAsync(ulong bytesConsumed)
             {
-                ValidateCopyToArgs(this, destination, bufferSize);
+                while (true)
+                {
+                    _connection._allowedReadLineBytes = MaxTrailingHeaderLength;
+                    if (LineIsEmpty(await _connection.ReadNextLineAsync().ConfigureAwait(false)))
+                    {
+                        break;
+                    }
+                }
+            }
 
-                return cancellationToken.IsCancellationRequested ? Task.FromCanceled(cancellationToken) :
-                    _connection != null ? CopyToAsyncCore(destination, bufferSize, cancellationToken) :
-                    Task.CompletedTask;
+            private void Finish()
+            {
+                _connection.ReturnConnectionToPool();
+                _connection = null;
             }
 
-            private async Task CopyToAsyncCore(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            private int ReadChunksFromConnectionBuffer(Span<byte> destination)
             {
-                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
-                try
+                int totalBytesRead = 0;
+                while (destination.Length > 0)
                 {
-                    if (_chunkBytesRemaining > 0)
+                    ReadOnlyMemory<byte> bytesRead = ReadChunkFromConnectionBuffer(destination.Length);
+                    Debug.Assert(bytesRead.Length <= destination.Length);
+                    if (bytesRead.Length == 0)
                     {
-                        await _connection.CopyToAsync(destination, _chunkBytesRemaining).ConfigureAwait(false);
-                        await ConsumeChunkBytesAsync(_chunkBytesRemaining).ConfigureAwait(false);
+                        break;
                     }
 
-                    while (await TryGetNextChunkAsync().ConfigureAwait(false))
-                    {
-                        await _connection.CopyToAsync(destination, _chunkBytesRemaining).ConfigureAwait(false);
-                        await ConsumeChunkBytesAsync(_chunkBytesRemaining).ConfigureAwait(false);
-                    }
+                    totalBytesRead += bytesRead.Length;
+                    bytesRead.Span.CopyTo(destination);
+                    destination = destination.Slice(bytesRead.Length);
                 }
-                catch (Exception exc) when (ShouldWrapInOperationCanceledException(exc, cancellationToken))
+                return totalBytesRead;
+            }
+
+            private ReadOnlyMemory<byte> ReadChunkFromConnectionBuffer(int maxBytesToRead)
+            {
+                ReadOnlySpan<byte> currentLine;
+                switch (_state)
                 {
-                    throw CreateOperationCanceledException(exc, cancellationToken);
+                    case ParsingState.ExpectChunkHeader:
+                        // Read the chunk header line.
```

Comment:
Should we assert that _chunkBytesRemaining = 0?

Reply:
Same for ExpectChunkTerminator case

Reply:
Sure